_posts/2025-11-06-sealing.markdown

```diff
@@ -0,0 +1,202 @@
+More on this later.
+
+If you couple a sealing key type with a language-level type (or family of self-disambiguating types), then you can build type safety that works in the presence of an attacker.
+It's common in C/C++ to expose an *opaque type* at API boundaries.
+This is usually implemented as a pointer to forward-declared structure type.
+C and C++ will not let you dereference this pointer unless you cast it to another type.
+CHERI sealing lets you enforce the no-dereference rule, even if malicious code does cast it to some other type.
+This means that you can hand sealed capabilities to other compartments and they must treat them as opaque types.
+When you get them back, you have a lightweight check that they really are the type that you expect.
+
+After the 2015 paper, we separated the permit-seal and permit-unseal permissions.
+In most CHERIoT use cases, the entity that can seal and unseal pointers with a particular otype is the same.
+This isn't universal.
+C++ provided one of our original use cases for separating them.
+If you seal C++ vtables with a well-known otype, and make the permit-unseal capability for it available anywhere, then you can have a C++ ABI where only the loader can forge vtables.
+This makes code reuse attacks harder.
+
+There are other situations where an object type is used for *integrity* but not *confidentiality*.
+It is an attestation that some software trusted to seal with a particular otype has done so, even if anyone is then able to unseal the result.
+The sentry mechanism (discussed later) is a variant of this idea.
+
+# Virtualised sealing shares hardware object types
+
+Morello had a 15-bit object-type field, which is sufficient for a lot of things.
+When we scaled CHERI down to 32-bit systems for CHERIoT, we ended up with only three bits of space.
+Of these, the zero value means not-sealed, so there were only seven values available.
+Even simple embedded software typically has more than seven types.
+CHERIoT systems usually have more than seven compartments, so seven types isn't even enough for one type per compartment (and some compartments wish to offer more than one sealed type).
+
+The *virtualised sealing* mechanism was designed to work around this limitation.
+We reserve two of the object types for objects that are instances of a structure with the following layout:
+
+```c
+struct SealedObject
+{
+    /// The sealing type for this object.
+    uint32_t type;
+    /// Padding for alignment
+    uint32_t padding;
+    /// The real data for this.
+    char data[];
+};
+```
+
+The first 32-bit word is the type, followed by 32 bits of padding, and then the real object.
+
+The hardware only uses values less than 16 for object types.
+Executable and data capabilities have different hardware sealing namespaces, so 3 bits actually gives 14 sealing types, seven for data and seven for sentries (see later).
```

Comment:
This paragraph feels out of place.  Maybe move it to before the introduction of virtualized sealing?